src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+            NormalizedSnapshotSpanCollection selectedSpans, ValueTuple command, CancellationToken cancellationToken)
+        {
+            var emptyResult = new CommentSelectionResult(new List<TextChange>(), new List<CommentTrackingSpan>(), Operation.Uncomment);
+
+            var experimentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleBlockComment))
+            {
+                return emptyResult;
+            }
+
+            var commentInfo = await service.GetInfoAsync(document, selectedSpans.First().Span.ToTextSpan(), cancellationToken).ConfigureAwait(false);
+            if (commentInfo.SupportsBlockComment)
+            {
+                return await ToggleBlockComments(document, commentInfo, selectedSpans, cancellationToken).ConfigureAwait(false);
+            }
+
+            return emptyResult;
+        }
+
+        /// <summary>
+        /// Retrieves data about the commented selection that can be optionally overriden by subclasses.
+        /// </summary>
+        protected virtual async Task<IToggleBlockCommentDocumentDataProvider> GetBlockCommentDocumentData(Document document, ITextSnapshot snapshot,
+            CommentSelectionInfo commentInfo, CancellationToken cancellationToken)
+        {
+            return new ToggleBlockCommentDocumentDataProvider(snapshot, commentInfo);
+        }
+
+        private async Task<CommentSelectionResult> ToggleBlockComments(Document document, CommentSelectionInfo commentInfo,
+            NormalizedSnapshotSpanCollection selectedSpans, CancellationToken cancellationToken)
+        {
+            var blockCommentDataProvider = await GetBlockCommentDocumentData(document, selectedSpans.First().Snapshot, commentInfo, cancellationToken).ConfigureAwait(false);
+
+            var blockCommentedSpans = blockCommentDataProvider.GetBlockCommentsInDocument();
+            var blockCommentSelections = selectedSpans.Select(span => new BlockCommentSelectionHelper(blockCommentedSpans, span)).ToList();
+
+            var returnOperation = Operation.Uncomment;
+
+            var uncommentChanges = new List<TextChange>();
+            var uncommentTrackingSpans = new List<CommentTrackingSpan>();
+            // Try to uncomment until an uncommented span is found.
+            foreach (var blockCommentSelection in blockCommentSelections)
+            {
+                // If any selection does not have comments to remove, then the operation should be comment.
+                if (!TryUncommentBlockComment(blockCommentedSpans, blockCommentSelection, uncommentChanges, uncommentTrackingSpans, commentInfo))
+                {
+                    returnOperation = Operation.Comment;
+                    break;
+                }
+            }
+
+            if (returnOperation == Operation.Comment)
```

Comment:
this only happens if we set the value in the loop above.  So why not just inline this into that case.  there's no other way to reach this.

Reply:
I'd rather not inline it just because in the operation.comment case, I'd be iterating through the same list as in the loop above.  I just stylistically prefer not having that second loop be inside the first loop.  If there's a strong preference the other way though I can change it.

Reply:
i can see the concern.  i'm not a fan of the curent flow control (have a loop, only to break out, only to have code that handles that case you found in the loop).  Two options:

1. you keep it (since i don't hate it).
2. you consider some other construct, like a local-function or helper-method here.

Reply:
Up to you.